### PR TITLE
BUG: fix unravel_index when dimension is greater than 'intp'

### LINKED ISSUE #9538: unravel_index: ValueError with larger integers
```
In [23]: numpy.unravel_index(1,[4294967295,2147483647])
Out[23]: (0, 1)

In [24]: numpy.unravel_index(1,[4294967295,2147483648])
Out[24]: (0, 1)

In [25]: numpy.unravel_index(1,[4294967295,2147483649])

ValueError                                Traceback (most recent call last)
<ipython-input-25-6551b793f30b> in <module>()
----> 1 numpy.unravel_index(1,[4294967295,2147483649])

ValueError: invalid entry in index array

In [26]: numpy.__version__
Out[26]: '1.13.1'`
```
Python 2.7.12, Ubuntu 14.04. Is this a known issue?

Thanks

### PR DESCRIPTION
**Pull request by psschand**
this PR fixes #9538

### CODE REVIEW

**Review by eric-wieser — COMMENTED** [review]
Does `PyArray_OverflowMultiplyList` set a useful error already?

**Comment by psschand** [comment]
@eric-wieser I also assumed that but it's not the case
when I ran it without setting the error, it showed 
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/akshay/workspace/numpy/env/lib/python3.7/site-packages/numpy/core/overrides.py", line 150, in public_api
    implementation, public_api, relevant_args, args, kwargs)
SystemError: <built-in function unravel_index> returned -1 without setting an error
```

**Comment by seberg** [comment]
@psschand the most important thing here will be adding a test for this. Also maybe take a look at https://docs.scipy.org/doc/numpy/dev/gitwash/development_workflow.html if you have not already done so.

Technically, `int64` is not correct, since this should be `intp`, which can be 32bits on 32bit platforms. It is a pretty backwater corner of numpy though, so I think you can just paraphrase it that the shape would lead to too many elements or so (would have to think about it a bit more).

**Comment by psschand** [comment]
@seberg I was just thinking.. doing that sounds more like a warning instead of error message right?

**Comment by psschand** [comment]
@seberg I've added tests can you check?

File: numpy/lib/tests/test_index_tricks.py [hunk 1]
```diff
@@ -106,6 +106,9 @@ def test_big_indices(self):
                 np.ravel_multi_index(arr, (41, 7, 120, 36, 2706, 8, 6)),
                 [5627771580, 117259570957])
 
+        # test unravel_index for big indices (issue #9538)
+        assert_raises(ValueError, np.unravel_index, 1, (4294967295, 2147483649))
```
Comment (eric-wieser) [inline numpy/lib/tests/test_index_tricks.py]: nit: the test intent is clearer as `(2**32-1, 2**31+1)`

**Review by eric-wieser — APPROVED** [review]

File: numpy/core/src/multiarray/compiled_base.c [hunk 2]
```diff
@@ -1261,7 +1261,12 @@ arr_unravel_index(PyObject *self, PyObject *args, PyObject *kwds)
         goto fail;
     }
 
-    unravel_size = PyArray_MultiplyList(dimensions.ptr, dimensions.len);
+    unravel_size = PyArray_OverflowMultiplyList(dimensions.ptr, dimensions.len);
+    if (unravel_size == -1) {
+        PyErr_SetString(PyExc_ValueError,
+                        "dimensions greater than 'intp' is not supported");
```
Comment (seberg) [inline numpy/core/src/multiarray/compiled_base.c]: Really just a thought:
```suggestion
                        "dimensions are too large; arrays and shapes with 
                        "a total size "greater than 'intp' are not supported.");
```

File: numpy/lib/tests/test_index_tricks.py (hunk 1 above)
Comment (seberg) [reply numpy/lib/tests/test_index_tricks.py]: On 32bits this ValueError is actually `ValueError: Maximum allowed dimension exceeded` a bit earlier on. Could get around by using 3 values, but does not really matter anyway.

**Review by seberg — APPROVED** [review]
Thanks @psschand, change things or not. Whichever you like. If you want to train your git-foo a bit, you could also squash the commits into a single one with a nice commit message (but we can also do that when merging into NumPy).

**Comment by psschand** [comment]
@seberg I made all the required changes.

**Comment by seberg** [comment]
@psschand sorry to be nagging, can you do Erics suggestion as well? If not, I can push the changes for you, but it does seem a bit nicer.

**Comment by psschand** [comment]
@seberg no problem, I've made it.

**Comment by seberg** [comment]
Thanks @psschand!

**Pull request merged**